### PR TITLE
Update RoiLabel class and add RoiImageFeatureToBatch

### PR DESCRIPTION
**Pull request by Menooker**
## What changes were proposed in this pull request?

Enabling masks info in RoiLabel class. Adding a new transformer to convert ImageFeatures with RoiLabels into mini-batches (RoiMiniBatch)

## How was this patch tested?

(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If it is possible, please attach a screenshot; otherwise, remove this)

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala [hunk 1]
```diff
@@ -25,26 +25,39 @@ import com.intel.analytics.bigdl.utils.{T, Table}
...
     require(classes.size(2) == bboxes.size(1), s"the number of classes ${ classes.size(2) }" +
       s"should be equal to the number of bounding box numbers ${ bboxes.size(1) }")
+    masks.foreach( m =>
+      require(classes.size(2) == m.length, s"the number of classes ${ classes.size(2) }" +
+        s"should be equal to the number of bounding box numbers ${m.length}")
+    )
   }
 
 
   def toTable: Table = {
```
Comment (Menooker) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala]: Not sure if we can change the format of the generated table. In the project, this `toTable` method is never used.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala [hunk 2]
```diff
@@ -25,26 +25,39 @@ import com.intel.analytics.bigdl.utils.{T, Table}
  * @param classes N (class labels) or 2 * N, the first row is class labels,
  * the second line is difficults
  * @param bboxes N * 4
+ * @param masks the array of annotation masks of the targets
  */
-case class RoiLabel(classes: Tensor[Float], bboxes: Tensor[Float]) {
+case class RoiLabel(classes: Tensor[Float], bboxes: Tensor[Float],
+  masks: Option[Array[Tensor[Byte]]] = None) {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala]: We usually use null by default for other definition, please remove `Option`

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala [hunk 3]
```diff
@@ -25,26 +25,39 @@ import com.intel.analytics.bigdl.utils.{T, Table}
...
   def copy(target: RoiLabel): Unit = {
     classes.resizeAs(target.classes).copy(target.classes)
     bboxes.resizeAs(target.bboxes).copy(target.bboxes)
+    require(!target.masks.isDefined, "Do not support support copying RoiLabels with masks")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala]: RoiLabels with masks not supported.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala [hunk 4]
```diff
@@ -57,11 +58,9 @@ class MTImageFeatureToBatch private[bigdl](width: Int, height: Int,
     _ => new PreFetch -> transformer.cloneTransformer()
   ).toArray
 
-  private val frameLength = height * width
-  private val featureData: Array[Float] = new Array[Float](batchSize * frameLength * 3)
-  private val labelData: Array[Float] = new Array[Float](batchSize)
-  private val featureTensor: Tensor[Float] = Tensor[Float]()
-  private val labelTensor: Tensor[Float] = Tensor[Float]()
+  protected def onEachImageFeature(img: ImageFeature, position: Int)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala]: `on` more for event triggering. how about `processImageFeature` ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala [hunk 5]
```diff
@@ -57,11 +58,9 @@ class MTImageFeatureToBatch private[bigdl](width: Int, height: Int,
...
+
+  protected def onEachBatch(batchSize: Int): MiniBatch[Float]
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala]: same as above, name it `createBatch` ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala [hunk 6]
```diff
@@ -127,3 +118,38 @@ private class PreFetch extends Transformer[ImageFeature, ImageFeature] {
...
+
+/**
+ * A transformer pipeline wrapper to create Minibatch in multiple threads
+ * @param width final image width
+ * @param height final image height
+ * @param totalBatchSize global batch size
+ * @param transformer pipeline for pre-processing
+ * @param toRGB  if converted to RGB, default format is BGR
+ */
+class MTImageFeatureToBatch private[bigdl](width: Int, height: Int,
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala]: `MT` means `Multi threading`, I'd like to have the abstract class to have this name,

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala]: and name it to another. To keep backward compatibility, Probably we need to have `MTImageFeatureToBatch ` object to instantiate different implementation with options

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala [hunk 7]
```diff
@@ -153,3 +157,37 @@ class MTImageFeatureToBatch private[bigdl](width: Int, height: Int,
...
+class RoiMTImageFeatureToBatch private[bigdl](width: Int, height: Int,
+  totalBatchSize: Int, transformer: FeatureTransformer, toRGB: Boolean = true)
+  extends MTImageFeatureToBatch(totalBatchSize, transformer) {
+
+  private val frameLength = height * width
+  private val featureData: Array[Float] = new Array[Float](batchSize * frameLength * 3)
+  private val labelData: Array[RoiLabel] = new Array[RoiLabel](batchSize)
+  private var featureTensor: Tensor[Float] = null
+
+
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala]: remove one empty line ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala [hunk 8]
```diff
@@ -28,33 +28,35 @@ import com.intel.analytics.bigdl.utils.{T, Table}
  * @param masks the array of annotation masks of the targets
  */
 case class RoiLabel(classes: Tensor[Float], bboxes: Tensor[Float],
-  masks: Option[Array[Tensor[Byte]]] = None) {
+  masks: Array[Tensor[Byte]] = null) {
   def copy(target: RoiLabel): Unit = {
     classes.resizeAs(target.classes).copy(target.classes)
     bboxes.resizeAs(target.bboxes).copy(target.bboxes)
-    require(!target.masks.isDefined, "Do not support support copying RoiLabels with masks")
+    require(target.masks == null, "Do not support support copying RoiLabels with masks")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala]: => "RoiLabels with masks not supported" ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala [hunk 9]
```diff
@@ -16,32 +16,36 @@
...
 import com.intel.analytics.bigdl.tensor.{Storage, Tensor}
+import com.intel.analytics.bigdl.transform.vision.image.label.roi.RoiLabel
 import com.intel.analytics.bigdl.utils.Engine
-
 import scala.reflect.ClassTag
 
 object MTImageFeatureToBatch {
   def apply(width: Int, height: Int, batchSize: Int,
-            transformer: FeatureTransformer, toRGB: Boolean = true)
+            transformer: FeatureTransformer, toRGB: Boolean = true, isRoi: Boolean = false)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/MTImageFeatureToBatch.scala]: probably `extractRoi` is more meaningful ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala (hunk 8 above)
Comment (Menooker) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala]: Yes, the mask array is `val`, so we cannot change the size of the array. What if the user wants to copy the mask with a different size?

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala]: No,  I meant to update the log to above one

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/RoiMiniBatch.scala [hunk 10]
```diff
@@ -0,0 +1,52 @@
+/*
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/RoiMiniBatch.scala]: I prefer to push this class into the same file as `ClassificationMTImageFeatureToBatch` to keep hierarchy consistency.

**Comment by wzhongyuan** [comment]
LGTM

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala [hunk 11]
```diff
@@ -74,6 +75,22 @@ object RoiLabel {
   // ISCROWD and ORIGSIZE are stored in ImageFeature
   val ISCROWD = "is_crowd"
   val ORIGSIZE = "size"
+
+
+  def getClassesFromImgFeature(imf: ImageFeature): Tensor[Float] = imf[Tensor[Float]](CLASSES)
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/transform/vision/image/label/roi/RoiLabel.scala]: just name `getClasses`, `getBBox` ?

**Comment by Menooker** [comment]
Jenkins pass http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/909/

**Pull request merged**